src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/BackupsService.java

```diff
@@ -0,0 +1,190 @@
+import org.niis.xroad.restapi.exceptions.ErrorDeviation;
+import org.niis.xroad.restapi.openapi.model.Backup;
+import org.niis.xroad.restapi.repository.BackupsRepository;
+import org.niis.xroad.restapi.util.FormatUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+
+import java.io.File;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Backups service.
+ */
+@Slf4j
+@Service
+@PreAuthorize("isAuthenticated()")
+public class BackupsService {
+
+    private static final String BACKUP_GENERATION_FAILED = "backup_generation_failed";
+
+    private final BackupsRepository backupsRepository;
+    private final BackupsConverter backupsConverter;
+    private final ServerConfService serverConfService;
+    private final ExternalProcessRunner externalProcessRunner;
+
+    @Setter
+    private String generateBackupScriptPath;
+    private static final String BACKUP_FILENAME_DATE_TIME_FORMAT = "yyyyMMdd-HHmmss";
+
+    /**
+     * BackupsService constructor
+     * @param backupsRepository
+     */
+    @Autowired
+    public BackupsService(BackupsRepository backupsRepository, BackupsConverter backupsConverter,
+                          ServerConfService serverConfService, ExternalProcessRunner externalProcessRunner,
+                          @Value("${script.generate-backup.path}") String generateBackupScriptPath) {
+        this.backupsRepository = backupsRepository;
+        this.backupsConverter = backupsConverter;
+        this.serverConfService = serverConfService;
+        this.externalProcessRunner = externalProcessRunner;
+        this.generateBackupScriptPath = generateBackupScriptPath;
+    }
+
+    /**
+     * Return a list of available backup files
+     * @return
+     */
+    public List<Backup> getBackupFiles() {
+        List<File> backupFiles = backupsRepository.getBackupFiles();
+        List<Backup> backups = backupsConverter.convert(backupFiles);
+        setCreatedAt(backups);
+        return backups;
+    }
+
+    /**
+     * Delete a backup file or throws an exception if the file does not exist
+     * @param filename
+     * @throws BackupFileNotFoundException
+     */
+    public void deleteBackup(String filename) throws BackupFileNotFoundException {
+        if (!backupExists(filename).isPresent()) {
+            throw new BackupFileNotFoundException(getFileNotFoundExceptionMessage(filename));
+        }
+        backupsRepository.deleteBackupFile(filename);
+    }
+
+    /**
+     * Read backup file's content
+     * @param filename
+     * @return
+     * @throws BackupFileNotFoundException
+     */
+    public byte[] readBackupFile(String filename) throws BackupFileNotFoundException {
```

Comment:
How does this work with weird / hostile filenames? Could you use e.g. path travelsals ("../../etc/passwd")? 

Spring firewall prevents some of these on the controller layer currently, but service should be safe anyway.

Reply:
Before trying to read the file, the given filename is validated against existing backup files. If the given filename does not match to an existing backup filename, `BackupFileNotFoundException` exception is thrown. The existing backup file names must meet the requirements of the backup filename filter.

https://github.com/nordic-institute/X-Road/pull/385/files#diff-de66d07ae90cf7c796f633208bd7dc09R55

This should prevent accessing other than valid backup files.

Reply:
Good, I did not fully track how readBackupFile is used - seems exploits are not possible. Also, the added more strict filename validation helps in this.